File: test_c/test_write.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#include "iolayer.h"
#include "simulations.h"
#include "aux.h"

struct generic_record_t simulate_keylist_record(struct directory_t* pdir) {
    struct generic_record_t record;

    return record;
}

struct test_t {
    uint32_t x;
    uint32_t y;
    uint32_t z;
};

struct test_record_t {
    struct rkey_t key;
    struct test_t test;
};

void serialize_test(char **dest, struct test_t * test) {
    put_u32(dest, test->x);
    put_u32(dest, test->y);
    put_u32(dest, test->z);
}

uint32_t size_test(struct test_t *test) {
    return 12;
}

struct generic_record_t simulate_test_record(struct llio_t *llio, struct rdirectory_t *dir) {
    // initialize the test object
    struct generic_record_t record;
    struct test_t test;
    test.x = 1;
    test.y = 2;
    test.z = 3;
    char b[12]; char *buffer = b;
    serialize_test(&buffer, &test);
    record.blob = malloc(12);
    memcpy(record.blob, buffer, 12);

    // generate the key
    struct rstring_t class_name, obj_name, obj_title;
    ctor_nomemcopy_pstring(&class_name, "test_t", 6);
    ctor_nomemcopy_pstring(&obj_name, "test obj name", 13);
    ctor_nomemcopy_pstring(&obj_title, "test obj title", 14);
    ctor_withnames_key(&record.key, 
                       &class_name,
                       &obj_name,
                       &obj_title);
    record.key.seek_pdir = dir->seek_dir;
    record.key.key_bytes = size_key(&record.key);
    record.key.obj_bytes = size_test(&test);
    record.key.total_bytes = record.key.obj_bytes = record.key.key_bytes;

    return record;
}

struct keys_list_record_t simulate_keys_list_record_for_dir(struct llio_t *llio, 
                                                            struct rdirectory_t *dir,
                                                            struct generic_record_t *rec) {
    // init
    struct keys_list_record_t record; 
    record.length = 1;
    record.pkeys = &rec->key;

    ctor_key(&record.key);
    record.key.seek_pdir = dir->seek_dir;
    record.key.key_bytes = size_key(&record.key);
    record.key.obj_bytes = size_key(&rec->key) + 4;
    record.key.total_bytes = record.key.obj_bytes + record.key.key_bytes;

    return record;
}

int main(int argc, char** argv) {
    if (argc==1) {
        printf("no input file provided\n");
        return 0;
    }

    char *filename = argv[1];
    printf("filename = %s\n", filename);

    struct llio_t llio = open_to_write(filename);

    printf("location to write = %lu and should be %d\n",
           llio.location, 100 + llio.top_dir_rec.key.total_bytes);

    // we simulate the streamer and free segments records
    simulate_streamer_record(&llio);
    simulate_free_segments_record(&llio);

    printf("--- simulated streamer, free segments and top dir keys list records ---\n");

    print_key(&llio.streamer_record.key);
    printf("- - - - - - - - - - - - - -\n");
    print_key(&llio.free_segments_record.key);
    printf("- - - - - - - - - - - - - -\n");

    struct generic_record_t test_record = simulate_test_record(&llio, 
        &llio.top_dir_rec.dir);
    struct keys_list_record_t keys_list_record = simulate_keys_list_record_for_dir(
        &llio, &llio.top_dir_rec.dir, &test_record);
    printf("- - - - - - - - - - - - - -\n");
    printf("print test record key\n");
    printf("- - - - - - - - - - - - - -\n");
    print_key(&test_record.key);
    printf("- - - - - - - - - - - - - -\n");
    printf("print keys list key\n");
    printf("- - - - - - - - - - - - - -\n");
    print_key(&keys_list_record.key);
    printf("- - - - - - - - - - - - - -\n");
    write_generic_record(&llio, &test_record);

    // write the keys list for the root/top dir
    write_keys_list_record_for_dir(&llio, &keys_list_record, &llio.top_dir_rec.dir);
    printf("location to write = %lu and should be %d\n",
           llio.location, 100 + llio.top_dir_rec.key.total_bytes + 
           keys_list_record.key.total_bytes + test_record.key.total_bytes);
    
    close_from_write(&llio);
    
    print_key(&keys_list_record.key);

    return 0;
}

File: test_rust/src/bin/test_write.rs
extern crate libc;
extern crate bytebuffer;
extern crate test_rust;

use std::ffi::CString;
use std::ffi::CStr;
use std::os::raw::c_char;
use std::slice;
use std::env;

use test_rust::bootstrap::*;
use test_rust::iolayer::*;

fn main() {
    println!("hello world!");

    let args: Vec<String> = env::args().collect();

    // create an llio_t and close
    let fname = args[1].clone();
    let filename = CString::new(fname).unwrap();
    let mut llio = open_to_write(&filename);
    println!("{:?}", llio);

    // simulate streamer and free segments record
    simulate_streamer_record(&mut llio);
    simulate_free_segments_record(&mut llio);

    // simulate a test record
    let mut test_record = simulate_test_record(&llio.top_dir_rec.dir);

    // simulate keys list record
    let mut keys_list_record = simulate_keys_list_record(&llio.top_dir_rec.dir, &mut test_record);

    // write record
    write_generic_record(&mut llio, &mut test_record);
    let mut dir = llio.top_dir_rec.dir;
    write_keys_list_record_for_dir(&mut llio,
                                   &mut keys_list_record,
                                   &mut dir);
    // TODO: we must coy and then reassign as we can not take 2 mutable references to the llio
    // object
    llio.top_dir_rec.dir = dir;

    // write required records
    close_from_write(llio);
}

trait Serializeable {
    fn serialize(&self) -> bytebuffer::ByteBuffer;
}

struct Test {
    x: i32,
    y: i32,
    z: i32
}

impl Serializeable for Test {
    fn serialize(&self) -> bytebuffer::ByteBuffer {
        let mut buffer = bytebuffer::ByteBuffer::new();
        buffer.write_i32(self.x);
        buffer.write_i32(self.y);
        buffer.write_i32(self.z);
        buffer
    }
}

fn simulate_keys_list_record(dir: &directory_t, gen_record: &mut generic_record_t)
    -> keys_list_record_t
{
    let class_name = CString::new("KeysList").unwrap();
    let obj_name = CString::new("").unwrap();
    let obj_title = CString::new("").unwrap();
    let mut key = key_t {
        object: object_t {version: 1, id: 1, bits: 1},
        total_bytes: 0,
        version: 1,
        obj_bytes: 0,
        date_time: datime_t {object: object_t {version: 1, id: 1, bits: 1}, raw: 12345},
        key_bytes: 0,
        cycle: 1,
        seek_key: 0,
        seek_pdir: 0,
        class_name: string_t {size: 8, cstr: class_name.as_ptr()},
        obj_name: string_t {size: 0, cstr: obj_name.as_ptr()},
        obj_title: string_t {size: 0, cstr: obj_title.as_ptr()}
    };
    key.seek_pdir = dir.seek_dir;
    key.key_bytes = size_key(&key) as u16;
    key.obj_bytes = size_key(&gen_record.key) + 4;
    key.total_bytes = key.obj_bytes + (key.key_bytes as u32);

    let record = keys_list_record_t {
        key: key,
        length: 1,
        pkeys: (&mut gen_record.key) as *mut key_t
    };

    std::mem::forget(class_name);
    std::mem::forget(obj_name);
    std::mem::forget(obj_title);
    return record
}

fn simulate_test_record(dir: &directory_t) 
    -> generic_record_t 
{
    // create a test object and serialize it
    let test = Test {
        x: 1,
        y: 2,
        z: 3
    };
    let mut blob = test.serialize().to_bytes();
    
    // generate a key for the test object
    let class_name = CString::new("Test").unwrap();
    let obj_name = CString::new("test obj name").unwrap();
    let obj_title = CString::new("test obj title").unwrap();
    let mut key = key_t {
        object: object_t {version: 1, id: 1, bits: 1},
        total_bytes: 0,
        version: 1,
        obj_bytes: 0,
        date_time: datime_t {object: object_t {version: 1, id: 1, bits: 1}, raw: 12345},
        key_bytes: 0,
        cycle: 1,
        seek_key: 0,
        seek_pdir: 0,
        class_name: string_t {size: 4, cstr: class_name.as_ptr()},
        obj_name: string_t {size: 13, cstr: obj_name.as_ptr()},
        obj_title: string_t {size: 14, cstr: obj_title.as_ptr()}
    };
    key.seek_pdir = dir.seek_dir;
    key.key_bytes = size_key(&key) as u16;
    key.obj_bytes = 12;
    key.total_bytes = key.obj_bytes + (key.key_bytes as u32);

    let record = generic_record_t {
        key: key,
        blob: (blob.as_mut_ptr()) as *mut i8
    };

    // need these guys to make rust not to clean up the memory
    std::mem::forget(class_name);
    std::mem::forget(obj_name);
    std::mem::forget(obj_title);
    std::mem::forget(blob);
    record
}
